File: src/main/java/org/railwaystations/rsapi/core/ports/in/ManageProfileUseCase.java
package org.railwaystations.rsapi.core.ports.in;

import org.railwaystations.rsapi.core.model.User;

import java.util.Locale;
import java.util.Optional;

public interface ManageProfileUseCase {

    void changePassword(User user, String newPassword);

    void resetPassword(String nameOrEmail, String clientInfo);

    void register(User newUser, String clientInfo) throws ProfileConflictException;

    void updateProfile(User user, User newProfile, String clientInfo) throws ProfileConflictException;

    void resendEmailVerification(User user);

    Optional<User> emailVerification(String token);

    void deleteProfile(User user, String userAgent);

    void updateLocale(User user, Locale locale);

    class ProfileConflictException extends RuntimeException {

        public ProfileConflictException() {
            super("Name or eMail is already taken");
        }

    }

}

File: src/main/kotlin/org/railwaystations/rsapi/core/ports/ManageProfileUseCase.kt
package org.railwaystations.rsapi.core.ports

import org.railwaystations.rsapi.core.model.User
import java.util.*

interface ManageProfileUseCase {
    fun changePassword(user: User, newPassword: String)

    fun resetPassword(nameOrEmail: String, clientInfo: String?)

    @Throws(ProfileConflictException::class)
    fun register(newUser: User, clientInfo: String?)

    @Throws(ProfileConflictException::class)
    fun updateProfile(user: User, newProfile: User, clientInfo: String?)

    fun resendEmailVerification(user: User)

    fun emailVerification(token: String): User?

    fun deleteProfile(user: User, userAgent: String?)

    fun updateLocale(user: User, locale: Locale)

    class ProfileConflictException : RuntimeException("Name or eMail is already taken")
}
